### PR TITLE
Fix Array Deserialization Panic

### PR DESCRIPTION
**Pull request by ggwpez**
## Description

Use `ArrayVec` - as suggested [here](https://github.com/arkworks-rs/algebra/issues/834#issuecomment-2178950139) - to remove an `unwrap` in the decoding of arrays.

closes: #834

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by burdges** [comment]
Just use arrayvec here, my code stupidly had a memopry leak if `T` allocates.  https://github.com/arkworks-rs/algebra/issues/834#issuecomment-2178950139

File: serialize/src/impls.rs [hunk 1]
```diff
@@ -458,13 +459,21 @@ impl<T: CanonicalDeserialize, const N: usize> CanonicalDeserialize for [T; N] {
...
+        }
         if let Validate::Yes = validate {
-            T::batch_check(result.iter())?
+            T::batch_check(array.iter())?
         }
-        Ok(result)
+        Ok(array
+            .into_inner()
+            // Defensive error return:
+            .map_err(|_| SerializationError::NotEnoughSpace)?)
```
Comment (mmagician) [inline serialize/src/impls.rs:476]: can this line ever throw an error? It looks like `NotEnoughSpace` should have been caught in the loop above

Comment (ggwpez) [reply serialize/src/impls.rs:476]: It should not, which is why i commented `Defensive`.  
I can also unwrap it if you prefer, but then `T: Debug` is required.

Comment (ggwpez) [reply serialize/src/impls.rs:476]: It [does](https://doc.rust-lang.org/std/result/enum.Result.html#method.expect)... but i can use  an `.unwrap_or_else(|_| unreachable!())`.

Comment (Pratyush) [reply serialize/src/impls.rs]: Oh I was looking at `Option::unwrap()` lol

Comment (burdges) [reply serialize/src/impls.rs]: `.map_err(|_| ()).unwrap()` works too.

**Review by Pratyush — APPROVED** [review]

**Review by mmagician — APPROVED** [review]

**Pull request merged**